desc/singularities.py

```diff
@@ -0,0 +1,733 @@
+        ), "singular integration requires source grid on a single surface"
+        assert src_grid.num_nodes == (
+            src_grid.num_theta * src_grid.num_zeta
+        ), "singular integration requires a tensor product grid in theta and zeta"
+        src_theta = src_grid.nodes[:, 1].reshape(
+            (src_grid.num_zeta, src_grid.num_theta)
+        )
+        src_zeta = src_grid.nodes[:, 2].reshape((src_grid.num_zeta, src_grid.num_theta))
+        assert isalmostequal(
+            src_theta, axis=0
+        ), "singular integration requires rectangular source grid in theta and zeta"
+        assert isalmostequal(
+            src_zeta, axis=1
+        ), "singular integration requires rectangular source grid in theta and zeta"
+        assert islinspaced(
+            src_theta, axis=1
+        ), "singular integration requires source nodes be equally spaced in theta"
+        assert islinspaced(
+            src_zeta, axis=0
+        ), "singular integration requires source nodes be equally spaced in zeta"
+
+        # need eval_grid to be linearly spaced in theta, zeta,
+        # and contain only 1 rho value
+        assert isalmostequal(
+            eval_grid.nodes[:, 0]
+        ), "singular integration requires eval grid on a single surface"
+        assert eval_grid.num_nodes == (
+            eval_grid.num_theta * eval_grid.num_zeta
+        ), "singular integration requires a tensor product grid in theta and zeta"
+        eval_theta = eval_grid.nodes[:, 1].reshape(
+            (eval_grid.num_zeta, eval_grid.num_theta)
+        )
+        eval_zeta = eval_grid.nodes[:, 2].reshape(
+            (eval_grid.num_zeta, eval_grid.num_theta)
+        )
+        assert isalmostequal(
+            eval_theta, axis=0
+        ), "singular integration requires rectangular eval grid in theta and zeta"
+        assert isalmostequal(
+            eval_zeta, axis=1
+        ), "singular integration requires rectangular eval grid in theta and zeta"
+        assert islinspaced(
+            eval_theta, axis=1
+        ), "singular integration requires eval nodes be equally spaced in theta"
+        assert islinspaced(
+            eval_zeta, axis=0
+        ), "singular integration requires eval nodes be equally spaced in zeta"
+
+        self._eval_grid = eval_grid
+        self._src_grid = src_grid
+        self._s = s
+        self._q = q
+
+        r, w, dr, dw = _get_quadrature_nodes(q)
+
+        src_dtheta = src_grid.spacing[:, 1]
+        src_dzeta = src_grid.spacing[:, 2] / src_grid.NFP
+
+        self._h_t = jnp.mean(src_dtheta)
+        self._h_z = jnp.mean(src_dzeta)
+
+        self._st = s / 2 * self._h_t * r * jnp.sin(w)
+        self._sz = s / 2 * self._h_z * r * jnp.cos(w)
+
+    def __call__(self, f, i):
+        """Interpolate data to polar grid points.
+
+        Parameters
+        ----------
+        f : ndarray
+            Data at source grid points to interpolate.
+        i : int
+            Index of polar node.
```

Comment:
How do these correspond to the `eval_grid`? Or are they separate?

Reply:
this corresponds to the `i`th polar node ($\rho_i$, $\omega_i$) around each ($\theta$, $\zeta$) point in eval grid